lms/services/email_unsubscribe.py

```diff
@@ -0,0 +1,63 @@
+from datetime import timedelta
+from functools import partial
+from typing import Callable
+
+from sqlalchemy import exists
+
+from lms.models import EmailUnsubscribe
+from lms.services.jwt import JWTService
+from lms.services.upsert import bulk_upsert
+
+
+class EmailUnsubscribeService:
+    def __init__(self, db, jwt_service: JWTService, secret: str, route_url: Callable):
+        self._db = db
+        self._jwt_service = jwt_service
+        self._secret = secret
+        self._route_url = route_url
+
+    def unsubscribe_url(self, email, tag):
+        """Generate the url for `email.unsubscribe` with the right token."""
+        token = self._generate_token(email, tag)
+        return self._route_url("email.unsubscribe", token=token)
+
+    def unsubscribe(self, token):
+        """Create a new entry in EmailUnsubscribe based on the email and tag encode in `token`."""
+        data = self._decode_token(token)
+
+        bulk_upsert(
```

Comment:
Not really `bulk` here but that function takes care of the upsert part without any extra conditionals here.